File: benchmark.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <string>
#include <chrono>
#include <algorithm>

using namespace std;
using namespace std::chrono;

const int MAXN = 1000005; // Larger limit for benchmarking
vector<int> adj[MAXN];
int low[MAXN], num[MAXN], cnt = 0;
bool is_articulation[MAXN];
int bridge_count = 0;
int n, m;

void dfs(int u, int p = 0) {
    num[u] = low[u] = ++cnt;
    int children = 0;
    for (int v : adj[u]) {
        if (v == p) continue;
        if (!num[v]) {
            children++;
            dfs(v, u);
            low[u] = min(low[u], low[v]);
            if (low[v] == num[v]) bridge_count++;
            if (p != 0 && low[v] >= num[u]) is_articulation[u] = true;
        } else {
            low[u] = min(low[u], num[v]);
        }
    }
    if (p == 0 && children > 1) is_articulation[u] = true;
}

void reset_state() {
    cnt = 0;
    bridge_count = 0;
    for (int i = 0; i <= n; i++) {
        num[i] = low[i] = 0;
        is_articulation[i] = false;
        adj[i].clear();
    }
}

int main(int argc, char* argv[]) {
    string suffix = (argc > 1) ? argv[1] : "1000";
    string filename = "datasets/dataset_" + suffix + ".txt";

    ifstream infile(filename);
    if (!infile) {
        cerr << "Error: File " << filename << " not found!" << endl;
        return 1;
    }

    // --- Loading Phase ---
    auto t1 = high_resolution_clock::now();
    infile >> n >> m;
    for (int i = 0; i < m; i++) {
        int u, v;
        infile >> u >> v;
        if (u < MAXN && v < MAXN) {
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
    }
    auto t2 = high_resolution_clock::now();

    // --- Algorithm Phase ---
    auto t3 = high_resolution_clock::now();
    for (int i = 1; i <= n; i++) {
        if (!num[i]) dfs(i);
    }
    int art_pts = 0;
    for (int i = 1; i <= n; i++) if (is_articulation[i]) art_pts++;
    auto t4 = high_resolution_clock::now();

    // --- Statistics ---
    auto load_time = duration_cast<milliseconds>(t2 - t1);
    auto algo_time = duration_cast<microseconds>(t4 - t3);

    cout << ">> Result for " << filename << endl;
    cout << "   Nodes: " << n << " | Edges: " << m << endl;
    cout << "   Load Time: " << load_time.count() << " ms" << endl;
    cout << "   Algo Time: " << algo_time.count() << " us (" << (double)algo_time.count()/1000.0 << " ms)" << endl;
    cout << "   Found: " << art_pts << " Articulation Points and " << bridge_count << " Bridges." << endl;
    cout << "-----------------------------------------------" << endl;

    return 0;
}

File: benchmark.py
import sys
import time
import os

# CRITICAL: Python's default limit is 1000. 
# We increase it to handle deep DFS trees in large datasets.
sys.setrecursionlimit(10**6)

def benchmark():
    # Get dataset suffix from command line argument
    suffix = sys.argv[1] if len(sys.argv) > 1 else "1000"
    file_path = f"datasets/dataset_{suffix}.txt"

    if not os.path.exists(file_path):
        print(f"Error: {file_path} not found.")
        return

    # --- Loading Phase ---
    start_load = time.time()
    with open(file_path, 'r') as f:
        lines = f.readlines()
        if not lines: return
        
        n, m = map(int, lines[0].split())
        adj = [[] for _ in range(n + 1)]
        for i in range(1, m + 1):
            u, v = map(int, lines[i].split())
            adj[u].append(v)
            adj[v].append(u)
    end_load = time.time()

    # --- Algorithm Phase ---
    num = [0] * (n + 1)
    low = [0] * (n + 1)
    is_articulation = [False] * (n + 1)
    timer = 0
    bridge_count = 0

    def dfs(u, p):
        nonlocal timer, bridge_count
        timer += 1
        num[u] = low[u] = timer
        children = 0
        
        for v in adj[u]:
            if v == p: continue
            if num[v] == 0:
                children += 1
                dfs(v, u)
                low[u] = min(low[u], low[v])
                
                if low[v] == num[v]:
                    bridge_count += 1
                if p != -1 and low[v] >= num[u]:
                    is_articulation[u] = True
            else:
                low[u] = min(low[u], num[v])
        return children

    start_algo = time.time()
    for i in range(1, n + 1):
        if num[i] == 0:
            if dfs(i, -1) > 1:
                is_articulation[i] = True
    end_algo = time.time()

    # --- Results ---
    load_ms = (end_load - start_load) * 1000
    algo_ms = (end_algo - start_algo) * 1000
    art_count = sum(is_articulation)

    print(f"--- Python Benchmark Results [{suffix}] ---")
    print(f"Nodes: {n} | Edges: {m}")
    print(f"Load Time: {load_ms:.2f} ms")
    print(f"Algo Time: {algo_ms:.2f} ms")
    print(f"Found: {art_count} Articulation Points, {bridge_count} Bridges")
    print("-" * 40)

if __name__ == "__main__":
    benchmark()
